spec/services/vine_voucher_validator_service_spec.rb

```diff
@@ -0,0 +1,365 @@
+            voucher_short_code: voucher_code,
+            voucher_set_id: "9d24349c-1fe8-4090-988b-d7355ed32559",
+            is_test: 1,
+            voucher_value_original: 500,
+            voucher_value_remaining: 500,
+            num_voucher_redemptions: 0,
+            last_redemption_at: "null",
+            created_at: "2024-10-01T13:20:02.000000Z",
+            updated_at: "2024-10-01T13:20:02.000000Z",
+            deleted_at: "null"
+          }
+        }.deep_stringify_keys
+      }
+
+      it "verifies the voucher with VINE API" do
+        expect(vine_api_service).to receive(:voucher_validation)
+          .and_return(mock_api_response( success: true, data:))
+
+        validate_voucher_service.validate
+      end
+
+      it "creates a new VINE voucher" do
+        allow(vine_api_service).to receive(:voucher_validation)
+          .and_return(mock_api_response( success: true, data:))
+
+        vine_voucher = validate_voucher_service.validate
+
+        expect(vine_voucher).not_to be_nil
+        expect(vine_voucher.code).to eq(voucher_code)
+        expect(vine_voucher.amount).to eq(5.00)
+        expect(vine_voucher.voucher_type).to eq("VINE")
+        expect(vine_voucher.external_voucher_id).to eq("9d2437c8-4559-4dda-802e-8d9c642a0c1d")
+        expect(vine_voucher.external_voucher_set_id).to eq(
+          "9d24349c-1fe8-4090-988b-d7355ed32559"
+        )
+      end
+    end
+
+    context "when distributor is not connected to VINE" do
+      it "returns nil" do
+        expect(validate_voucher_service.validate).to be_nil
+      end
+
+      it "doesn't call the VINE API" do
+        expect(vine_api_service).not_to receive(:voucher_validation)
+
+        validate_voucher_service.validate
+      end
+
+      it "adds an error message" do
+        validate_voucher_service.validate
+
+        expect(validate_voucher_service.errors).to include(
+          { vine_settings: "No Vine api settings for the given enterprise" }
+        )
+      end
+
+      it "doesn't creates a new VINE voucher" do
+        expect { validate_voucher_service.validate }.not_to change { Voucher.count }
+      end
+    end
+
+    context "when there is an API error" do
+      let!(:vine_connected_app) {
+        ConnectedApps::Vine.create(
+          enterprise: distributor, data: { api_key: "1234567", secret: "my_secret" }
+        )
+      }
+
+      before do
+        allow(vine_api_service).to receive(:voucher_validation).and_raise(Faraday::Error)
+      end
+
+      it "returns nil" do
+        expect(validate_voucher_service.validate).to be_nil
+      end
+
+      it "adds an error message" do
+        validate_voucher_service.validate
+
+        expect(validate_voucher_service.errors).to include(
+          { vine_api: "There was an error communicating with the API" }
+        )
+      end
+
+      it "doesn't creates a new VINE voucher" do
+        expect { validate_voucher_service.validate }.not_to change { Voucher.count }
+      end
```

Comment:
I think this test, and the `nil` test, could easily be shared_examples between the next few contexts.

Reply:
I don't like using shared_examples I find them more confusing then helpful. I should be able to improve this though

Reply:
> I find them more confusing then helpful

Fair point, me too!